src/main/java/core/basesyntax/operations/PurchaseOperation.java

```diff
@@ -0,0 +1,29 @@
+package core.basesyntax.operations;
+
+import core.basesyntax.db.Storage;
+import core.basesyntax.model.FruitTransaction;
+
+public class PurchaseOperation implements OperationHandler {
+    private final Storage storage;
+
+    public PurchaseOperation(Storage storage) {
+        this.storage = storage;
+    }
+
+    @Override
+    public void apply(FruitTransaction transaction) {
+        if (transaction == null) {
+            throw new IllegalArgumentException("Transaction cannot be null.");
+        }
+
+        int currentQuantity = storage.getQuantity(transaction.getFruit());
+        int newQuantity = currentQuantity - transaction.getQuantity();
+
+        if (newQuantity < 0) {
+            throw new RuntimeException("Insufficient stock for purchase of "
+                    + transaction.getFruit());
+        }
+
+        storage.updateFruit(transaction.getFruit(), newQuantity);
+    }
+}

```

Comment:
The `PurchaseOperation` class is correctly implementing the `OperationHandler` interface. However, according to the checklist, you need to check the result balance before saving it in the Storage to ensure it is positive. This is correctly done in lines 19-27. Just make sure that the exception message in lines 23-24 is informative enough to understand which fruit and how much quantity was attempted to be purchased, which caused the negative balance.